File: src-tauri/src/commands/worker.rs
//! Worker lifecycle IPC (TASK-006).
//!
//! Exposes only the *state* of the Python sidecar — never the session token.
//! The token is owned by the Rust core and stays internal (no IPC, no UI, no
//! logs). Later worker API calls are proxied through controlled Rust commands.

use tauri::State;

use crate::services::worker_manager::{WorkerManager, WorkerStateInfo};

/// Returns the current lifecycle snapshot of the Python sidecar.
#[tauri::command]
pub fn get_worker_state(manager: State<'_, WorkerManager>) -> WorkerStateInfo {
    manager.state_info()
}

/// `worker.restart() → WorkerStateInfo` — stop the sidecar, then start it
/// again with a fresh ephemeral port + session token.
///
/// Idempotent-safe: `stop` blocks until the supervisor has cleaned up and
/// `start` rejects when a worker is already running, so this is the only
/// restart path the UI should use.
#[tauri::command]
pub fn restart(manager: State<'_, WorkerManager>) -> WorkerStateInfo {
    manager.stop();
    if let Err(e) = manager.start() {
        log::error!("worker restart failed: {e}");
    }
    manager.state_info()
}

File: src/api/worker.ts
import { safeInvoke } from "@/api/invoke";

/** Lifecycle state of the Python sidecar (mirrors Rust `WorkerState`). */
export type WorkerState = "stopped" | "starting" | "ready" | "stopping" | "failed";

/** Snapshot exposed over IPC (never contains the session token). */
export type WorkerStateInfo = {
  state: WorkerState;
  pid: number | null;
  port: number | null;
  restarts: number;
  last_error: string | null;
};

/** Current lifecycle snapshot of the Python sidecar. */
export function getWorkerState(): Promise<WorkerStateInfo> {
  return safeInvoke("worker.get_worker_state");
}

/** Stop and restart the Python sidecar (fresh port + session token). */
export function restartWorker(): Promise<WorkerStateInfo> {
  return safeInvoke("worker.restart");
}
